core/src/main/java/com/linecorp/armeria/internal/client/DefaultClientRequestContext.java

```diff
@@ -849,4 +851,24 @@ private String toStringSlow(@Nullable Channel ch, @Nullable RequestLogAccess par
             return buf.toString();
         }
     }
+
+    @Override
+    public CompletableFuture<Void> disconnectWhenFinished() {
+        final Channel ch = channel();
+        if (ch == null || !ch.isActive()) {
+            setAdditionalRequestHeader(HttpHeaderNames.CONNECTION, HttpHeaderValues.CLOSE);
+            return UnmodifiableFuture.completedFuture(null);
```

Comment:
This seems incorrect to me. The future returned here has to be completed when 1) the connection is closed or 2) the request has failed. How about adding a listener to its `RequestLog` so we can wait until we have more information?